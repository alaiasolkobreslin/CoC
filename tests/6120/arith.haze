
(* Natural numbers *)
let Nat = forall N:Type, (forall _:N, (forall _:(forall _:N, N), N)) in

(* Some numbers *)
let zero = lambda N:Type -> (lambda x:N -> (lambda _:(forall _:N, N) -> x)) in
let one = lambda N:Type -> (lambda x:N -> (lambda f:(forall _:N, N) -> (f x))) in
let two = lambda N:Type -> (lambda x:N -> (lambda f:(forall _:N, N) -> (f (f x)))) in

(* Some arithmetic functions *)
let succ = lambda nat:Nat -> (lambda N:Type -> (lambda x:N -> 
  (lambda f:(forall _:N, N) -> (f (((nat N) x) f))))) in
let add = lambda n1:Nat -> (lambda n2:Nat -> (lambda N:Type -> (lambda x:N ->
  (lambda f:(forall _:N, N) ->  (((n1 N) (((n2 N) x) f)) f))))) in

Theorem zero_is_nat: Nat. Proof. zero.
Theorem one_is_nat: Nat. Proof. one.
Theorem two_is_nat: Nat. Proof. two.

Theorem succ_zero_is_nat: Nat. Proof. (succ zero).

(succ (succ zero))
