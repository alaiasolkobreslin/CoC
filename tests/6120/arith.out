arith.haze

Let: Nat beta reduced: (forall N:Type , (forall _:N , (forall _:(forall _:N , N) , N)))

Let: zero beta reduced: (fun N:Type -> (fun x:N -> (fun _:(forall _:N , N) -> x)))

Let: one beta reduced: (fun N:Type -> (fun x:N -> (fun f:(forall _:N , N) -> (f x))))

Let: two beta reduced: (fun N:Type -> (fun x:N -> (fun f:(forall _:N , N) -> (f (f x)))))

Let: succ beta reduced: (fun nat:(forall N:Type , (forall _:N , (forall _:(forall _:N , N) , N))) -> (fun N:Type -> (fun x:N -> (fun f:(forall _:N , N) -> (f (((nat N) x) f))))))

Let: add beta reduced: (fun n1:(forall N:Type , (forall _:N , (forall _:(forall _:N , N) , N))) -> (fun n2:(forall N:Type , (forall _:N , (forall _:(forall _:N , N) , N))) -> (fun N:Type -> (fun x:N -> (fun f:(forall _:N , N) -> (((n1 N) (((n2 N) x) f)) f))))))
Theorem: zero_is_nat beta reduced: (forall N:Type , (forall _:N , (forall _:(forall _:N , N) , N)))
Proof typed: (forall N:Type , (forall x:N , (forall _:(forall _:N , N) , N)))
Proof beta reduced: (fun N:Type -> (fun x:N -> (fun _:(forall _:N , N) -> x)))


Theorem: one_is_nat beta reduced: (forall N:Type , (forall _:N , (forall _:(forall _:N , N) , N)))
Proof typed: (forall N:Type , (forall x:N , (forall f:(forall _:N , N) , N)))
Proof beta reduced: (fun N:Type -> (fun x:N -> (fun f:(forall _:N , N) -> (f x))))


Theorem: two_is_nat beta reduced: (forall N:Type , (forall _:N , (forall _:(forall _:N , N) , N)))
Proof typed: (forall N:Type , (forall x:N , (forall f:(forall _:N , N) , N)))
Proof beta reduced: (fun N:Type -> (fun x:N -> (fun f:(forall _:N , N) -> (f (f x)))))


Theorem: succ_zero_is_nat beta reduced: (forall N:Type , (forall _:N , (forall _:(forall _:N , N) , N)))
Proof typed: (forall N:Type , (forall x:N , (forall f:(forall _:N , N) , N)))
Proof beta reduced: (fun N:Type -> (fun x:N -> (fun f:(forall _:N , N) -> (f x))))


(fun N:Type -> (fun x:N -> (fun f:(forall _:N , N) -> (f (f x)))))
Term typed: (forall N:Type , (forall x:N , (forall f:(forall _:N , N) , N)))

